### PR TITLE
KAFKA-7869: Refactor RocksDBConfigSetter API to separate DBOptions and CFOptions

### PR DESCRIPTION
**Pull request by dongjinleekr**
This is a draft implementation of [KAFKA-7869](https://issues.apache.org/jira/browse/KAFKA-7869). @guozhangwang Is this what you intended?

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by dongjinleekr** [comment]
As a note, this PR consists of two commits:

1. Refactor `RocksDBStore`: make `DBOptions`, `ColumnFamilyOptions` instances into instance variables.
2. Actual implementation: add new methods to the interface, and call them if there is a non-default column family in the rocksdb.

**Comment by dongjinleekr** [comment]
As a note, this PR consists of two commits:

1. Refactor `RocksDBStore`: make `DBOptions`, `ColumnFamilyOptions` instances into instance variables.
2. Actual implementation: add new methods to the interface, and call them if there is a non-default column family in the rocksdb.

Please have a look when you are free.

**Comment by dongjinleekr** [comment]
Retest this please.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java [hunk 1]
```diff
@@ -177,6 +126,137 @@ void openDB(final ProcessorContext context) {
...
+            if (prepareForBulkload) {
+                dbOptions.setMaxBackgroundFlushes(4);
+                columnFamilyOptions.setDisableAutoCompactions(true);
+                columnFamilyOptions.setLevel0FileNumCompactionTrigger(1 << 30);
+                columnFamilyOptions.setLevel0SlowdownWritesTrigger(1 << 30);
+                columnFamilyOptions.setLevel0StopWritesTrigger(1 << 30);
+            }
+
+            if (configSetter != null) {
+                configSetter.close(name, dbOptions, columnFamilyOptions);
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:256]: configSetter.close() will clean up any resources constructed in configSetter.setConfig, and should only be called in RocksDBStore#close()

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java [hunk 2]
```diff
@@ -177,6 +126,137 @@ void openDB(final ProcessorContext context) {
...
+
+            if (prepareForBulkload) {
+                userSpecifiedOptions.prepareForBulkLoad();
+            }
+
+            dbOptions = new DBOptions(tempDbOptions);
+            columnFamilyOptions = new ColumnFamilyOptions(tempColumnFamilyOptions);
+
+            if (configSetter != null) {
+                configSetter.close(name, userSpecifiedOptions);
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:196]: ditto here, should be moved to RocksDBStore#close

**Comment by ableegoldman** [comment]
Thanks for the PR! One concern I have is here is that it may be non-obvious to users which version of ConfigSetter#setConfig (or ConfigSetter#close) should be implemented, or will be called. What we wouldn't want is for a user's implemented RocksDBConfigSetter to suddenly ignore their custom options because we are now calling the empty default of the new version of setConfig.
That said I believe this will require a KIP anyway, so maybe that's the best place to facilitate discussion of this?

**Comment by mjsax** [comment]
As mentioned on the Jira, I am still not convinced that we should split it.

**Comment by mjsax** [comment]
As discussed on the ticket, we think we should not split the class. Closing this PR.

**Pull request closed without merging**